events/events.go

```diff
@@ -0,0 +1,137 @@
+package events
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/flanksource/commons/logger"
+	"gorm.io/gorm"
+
+	"github.com/flanksource/incident-commander/api"
+	"github.com/flanksource/incident-commander/db"
+	responderPkg "github.com/flanksource/incident-commander/responder"
+)
+
+func ListenForEvents() {
+
+	logger.Infof("Started listening for events")
+
+	// Consume pending events
+	consumeEventsWrapper()
+
+	pgNotify := make(chan bool)
+	go listenToPostgresNotify(pgNotify)
+
+	for {
+		select {
+		case <-pgNotify:
+			consumeEventsWrapper()
+
+		case <-time.After(10 * time.Second):
+			logger.Debugf("timed out waiting for pgNotify")
+		}
+	}
+}
+
+func listenToPostgresNotify(pgNotify chan bool) {
+	ctx := context.Background()
+
+	conn, err := db.Pool.Acquire(ctx)
+	if err != nil {
+		logger.Errorf("Error creating database pool: %v", err)
+	}
+
+	_, err = conn.Exec(ctx, "LISTEN event_queue_updates")
+	if err != nil {
+		logger.Errorf("Error listening to database notify: %v", err)
+	}
+
+	for {
+		_, err = conn.Conn().WaitForNotification(ctx)
+		if err != nil {
+			logger.Errorf("Error waiting for database notifications: %v", err)
+		}
+
+		pgNotify <- true
+	}
+
+}
+
+// TODO: Better function name
+func consumeEventsWrapper() {
+	// Keep on iterating till the queue is empty
+	for {
+		err := consumeEvents()
+		if err != nil {
+			if errors.Is(err, gorm.ErrRecordNotFound) {
+				return
+			} else {
+				logger.Errorf("Error processing event: %v", err)
+			}
+		}
+	}
+}
+
+func consumeEvents() error {
+	var event api.Event
+
+	tx := db.Gorm.Begin()
+
+	// TODO: Add attempts where clause
+	err := tx.Raw("SELECT id, name, properties FROM event_queue FOR UPDATE SKIP LOCKED LIMIT 1").First(&event).Error
```

Comment:
In case of failed events, we are adding an error message and incrementing the attempt count.

But I feel we should not retry/pick an event which has >= N attempts. What should N be ? I was thinking 3 attempts are more than enough ?

Reply:
I think we should always retry, and then add an explicit action to flush failed attempts, the problem with stopping to to retry is the connection could have an expired token/password that takes a couple days to fix and by then you would need to perform some other manual action other than updating the password. 

We should be backing off, i.e. retry after 1, 5, 15, 60m, and there after every 3h